core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java

```diff
@@ -309,6 +309,10 @@ public IToolChain getToolChain() throws CoreException {
 		return toolChain;
 	}
 
+	protected void setToolChain(IToolChain toolChain) {
```

Comment:
The reason for my change is that I need support for multiple C compilers in one Makefile. The core build CMake flow also can do it. Only there is a cmake limitation to support only one compiler in one run.
We have discussed this in the cdt-dev mailing list "Managing include paths and macros in 'Core Build' projects." https://www.eclipse.org/lists/cdt-dev/threads.html#35316

I started with overriding getToolChain, and putting getToolChain everywhere. And this worked for me. At the end I added setToolChain, because I thought if anyone adds code that directly uses the toolChain variable my plugin will be broken.

If you are fine with adding setToolChain(), I will remove the other changes.

Reply:
I want to loop in @betamaxbandit because I think he has been trying to resolve a similar issue - see PR #75 

IIUC @betamaxbandit is expanding `IToolChain` (and others) to support a single "tool chain" that is actually multiple compilers.

This sounds like the same thing here.

When `IToolChain.getResourcesFromCommand` is called, then the tool chain should delegate to the right "sub-tool chain" (*something I just made up*). Maybe to make @betamaxbandit and your issue easier we need a "composite" tool chain that understands how to delegate properly. Perhaps something like:

```java
/**
 * @since 8.0
 */
public abstract class CompositeToolChain implements IToolChain {

	private List<IToolChain> children = new ArrayList<>();
	private IToolChainProvider provider;

	public CompositeToolChain(IToolChainProvider provider, List<IToolChain> children) {
		this.provider = provider;
		this.children.addAll(children);
	}

	@Override
	public IToolChainProvider getProvider() {
		return provider;
	}

	@Override
	public List<String> getBinaryParserIds() {
		return children.stream().flatMap(c -> c.getBinaryParserIds().stream()).collect(Collectors.toUnmodifiableList());
	}

	@Override
	public Path getCommandPath(Path command) {
		return children.stream().map(c -> c.getCommandPath(command)).filter(Objects::nonNull).findFirst().orElse(null);
	}

	@Override
	public String[] getCompileCommands() {
		return children.stream().flatMap(c -> Arrays.stream(c.getCompileCommands())).toArray(String[]::new);
	}

	@Override
	public IResource[] getResourcesFromCommand(List<String> command, URI buildDirectoryURI) {
		return children.stream().flatMap(c -> Arrays.stream(c.getResourcesFromCommand(command, buildDirectoryURI)))
				.toArray(IResource[]::new);
	}
}
```

To better understand the problem, what is the logic that your sub-class of `CBuildConfiguration` is using to change the tool chain? Can that logic exist in a composite tool chain as I described above?

Reply:
This is a case where the default compiler is gcc, and based on the command line I can change it to "myCompiler" using "myToolChain". I did this by overriding getToolChain(). Per processed line I set myToolChain.

`
    @Override
    public IToolChain getToolChain() throws CoreException {
        if (myToolChain == null) {
            return super.getToolChain();
        } else {
            return myToolChain;
        }
    }

    private IToolChain getToolChain(List<String> command) throws CoreException {
        if (command.size() > 0 && command.get(0).endsWith("/myCompiler")) {
            IToolChainManager toolChainManager = CCorePlugin.getService(IToolChainManager.class);
            Map<String, String> properties = new HashMap<String, String>() {
                private static final long serialVersionUID = 4493246628665145757L;

                {
                    put(IToolChain.ATTR_ARCH, "myArch");
                }
            };
            Collection<IToolChain> toolChains = toolChainManager.getToolChainsMatching(properties);
            if (toolChains.size() > 0) {
                myToolChain = toolChains.iterator().next();
                return myToolChain;
            } else {
                throw new CoreException(new Status(Status.ERROR, Activator.PLUGIN_ID,
                        "Failed to get toolchain for command " + command));
            }
        } else {
            myToolChain = null;
            return super.getToolChain();
        }
    }

    @Override
    public boolean processLine(String line, List<Job> jobsArray) {
        // Split line into args, taking into account quotes
        List<String> command = stripArgs(line);
        try {
            getToolChain(command);
        } catch (CoreException e1) {
            CCorePlugin.log(e1);
            return false;
        }
        return super.processLine(line, jobsArray);
    }

`

Reply:
FWIW this makes me double-down that change toolchain to non-final is a mistake. The code you have here leaves the toolchain in a non-deterministic state. So which toolchain `getBinaryParserId` uses will depend on what *happened* to be the last line processed in processLine.

---

OT but:

```java
            Map<String, String> properties = new HashMap<String, String>() {
            private static final long serialVersionUID = 4493246628665145757L;

            {
                put(IToolChain.ATTR_ARCH, "myArch");
            }
        };
```

can probably be replaced with:

```java
Map<String, String> properties = Map.of(IToolChain.ATTR_ARCH, "myArch");

Reply:
myToolChain is an extension of GCCToolChain, with an override of addDiscoveryOptions(). I also had to take a copy of the private method getScannerInfo(), because I had to change the matching in the "Macro reader" Thread.

Reply:
So, in your toolchain, can you do something like this?

```java
public class MyToolChain extends ContainerGCCToolChain {
	@Override
	public IResource[] getResourcesFromCommand(List<String> command, URI buildDirectoryURI) {
		if (command.size() > 0 && command.get(0).endsWith("/myCompiler")) {
			// do your processing here
			return your results; 
		}
		return super.getResourcesFromCommand(command, buildDirectoryURI);
	}
	
	@Override
	public List<String> stripCommand(List<String> command, IResource[] resources) {
		if (command.size() > 0 && command.get(0).endsWith("/myCompiler")) {
			// do your processing here
			return your results; 
		}
		return super.stripCommand(command, resources);
	}

//...
}
```

Reply:
For me it is not a problem, because in my case the binary parser id is always the same. But I understand there should be proper support for multiple bin parsers, before this can go into CDT.

Reply:
> For me it is not a problem

Is this in reference to the earlier comment  https://github.com/eclipse-cdt/cdt/pull/130#discussion_r1010401543? 

What about my last comment https://github.com/eclipse-cdt/cdt/pull/130#discussion_r1010552242? Does that resolve the issue, or is there something I am missing?

Reply:
> I want to loop in @betamaxbandit because I think he has been trying to resolve a similar issue - see PR #75
> 
> IIUC @betamaxbandit is expanding `IToolChain` (and others) to support a single "tool chain" that is actually multiple compilers.
> 
> This sounds like the same thing here.

Hi,
My usecase for requiring multiple binary parsers is similar to this, but I don't think quite the same. My project's CMake script builds using 2 cross toolchains which produce multiple binaries, each of which is loaded into a separate GDB at launch. You could argue this is a "composite" project, but I like to think of the 2 toolchains as separate because there are others scenarios where they are used individually.

Reply:
FYI
This problem looks similar to something I experienced in Sloeber.
I ended up creating multiple tools for the same source type; each with their own a name provider.
When the name provider returned null  that tool was not part of the makefile for this file.
This way I could change the tool(chain?) and options
I also needed extra info to make the decision in the name provider to make the correct decision.
I failed to contribute the extra info to CDT but the capability for the name provider to return null is part of CDT for some years now.

Reply:
> Is this in reference to the earlier comment [#130 (comment)](https://github.com/eclipse-cdt/cdt/pull/130#discussion_r1010401543)?
> 
> What about my last comment [#130 (comment)](https://github.com/eclipse-cdt/cdt/pull/130#discussion_r1010552242)? Does that resolve the issue, or is there something I am missing?

The CompositeToolChain looks like a good idea.

Reply:
> So, in your toolchain, can you do something like this?
> 

I had to completely override getResourcesFromCommand(), because in my command line the C source file can be in the middle, followed by other arguments without a dash. The implementation of GCCToolChain assumes the C source file is the last item in the line. I'm checking the file extension.

I can use stripCommand() unmodified.

Reply:
Here with check on extension.

```java
    @Override
    public IResource[] getResourcesFromCommand(List<String> cmd, URI buildDirectoryURI) {
        /*
         * Start at the back looking for arguments
         * GCCToolChain.getResourcesFromCommand() assumes the source files are at the
         * end of the command. In Volcano command line this is not the case.
         */
        List<IResource> resources = new ArrayList<>();
        IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
        for (int i = cmd.size() - 1; i >= 0; --i) {
            String arg = cmd.get(i);
            if (arg.startsWith("-")) { //$NON-NLS-1$
                // ran into an option, we're done.
                continue;
            }
            if (i > 1 && cmd.get(i - 1).equals("-o")) { //$NON-NLS-1$
                // this is an output file
                --i;
                continue;
            }
            if (arg.endsWith(".c")) {
                try {
                    Path srcPath = Paths.get(arg);
                    URI uri;
                    if (srcPath.isAbsolute()) {
                        uri = srcPath.toUri();
                    } else {
                        String mingwPath = fixMingwPath(arg);
                        if (mingwPath != arg) {
                            uri = Paths.get(mingwPath).toUri();
                        } else {
                            uri = Paths.get(buildDirectoryURI).resolve(srcPath).toUri().normalize();
                        }
                    }

                    for (IFile resource : root.findFilesForLocationURI(uri)) {
                        resources.add(resource);
                    }
                } catch (IllegalArgumentException e) {
                    // Bad URI
                    continue;
                }
                // For now we assume Volcano compiles one source file at a time.
                break;
            }
        }

        return resources.toArray(new IResource[resources.size()]);
    }
```

Reply:
MyToolChain.getResourcesFromCommand() is only called on lines which use myCompiler. So calling super is not needed.

Reply:
> The implementation of GCCToolChain assumes the C source file is the last item in the line.

That is a defect that should be resolved in CDT. Would you be able to provide your code as a PR without the `arg.endsWith(".c")` part?

That said, I don't know why this code didn't reuse\* `o.e.c.m.l.s.providers.AbstractBuildCommandParser.parseResourceName(String)` which handles all these cases and has many more years of testing under its belt (like `o.e.c.m.l.s.providers.tests.GCCBuildCommandParserTest`). My guess is that Doug was trying to simplify the code.

\* The reuse could be code or knowledge. Parsing a gcc command line is complicated and the `getResourcesFromCommand` in `GCCToolChain` is way too over-simplified.

Reply:
> That is a defect that should be resolved in CDT. Would you be able to provide your code as a PR without the `arg.endsWith(".c")` part?
> 
> That said, I don't know why this code didn't reuse* `o.e.c.m.l.s.providers.AbstractBuildCommandParser.parseResourceName(String)` which handles all these cases and has many more years of testing under its belt (like `o.e.c.m.l.s.providers.tests.GCCBuildCommandParserTest`). My guess is that Doug was trying to simplify the code.
> 
> * The reuse could be code or knowledge. Parsing a gcc command line is complicated and the `getResourcesFromCommand` in `GCCToolChain` is way too over-simplified.

OK. I will create a PR.